### ISSUE TITLE
PMML gives different predictions than Python/XGBoost

### ISSUE THREAD

**Issue by spencerhs5** [jpmml/sklearn2pmml#245]
Hi, 

I've narrowed this down (I think) to XGBoosts with PMML. It seemed to work okay with lightboost. 

The python generated PMML gives different predictions than the native python code. 

Code and data are in the zip file. The code is also printed below. 

Thanks in advance, let me know if you want anything else. 

```
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn2pmml.feature_extraction.text import Splitter
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.model_selection import GridSearchCV
from xgboost import XGBClassifier
from sklearn2pmml import sklearn2pmml
from sklearn2pmml import PMMLPipeline
from sklearn.feature_extraction import text
from lightgbm import LGBMClassifier



df = pd.read_csv('Trump.csv')
df.dropna(inplace=True)
train_ds, test_ds = train_test_split(df, test_size=0.3, random_state=42, stratify=df['Outcome'])

X = train_ds['Quote'].tolist()
y = train_ds['Outcome'].tolist()


vectorizer = TfidfVectorizer(
    analyzer='word',
    decode_error='ignore',
    tokenizer=Splitter(),
    stop_words="english",
    token_pattern=None,
    norm=None
)



textpipeline = PMMLPipeline([
    ("tf-idf", vectorizer),
    ('xgb', XGBClassifier())])



textpipeline.fit(X, y)


sklearn2pmml(textpipeline, "xgbtrumppredsvdefstop.pmml", debug='True')


test_X=test_ds['Quote'].tolist()
pred_Y=textpipeline.predict_proba(test_X)
#i see the first test record with p(0) =.588907 in python but .120747... from pmml
#first test record string: In the end, you're measured not by how much you undertake but by what you finally accomplish.
```

[TextTest.zip](https://github.com/jpmml/sklearn2pmml/files/5294971/TextTest.zip)

**Comment by vruusmann** [comment]
I'm fairly confident that JPMML is doing conversion and prediction correctly.

There must be something about your data and/or Python code. One thing that catches my eye is that some sentences in your CSV file start and end with quotes, whereas some others don't. That probably confuses the text tokenizer (eg. `I am` and `"I am` are treated as two different tokens in Python).

Will look into this a bit later to confirm the above hypothesis.

**Comment by spencerhs5** [comment]
Thanks-- I had noticed the start/end characters with a different dataset I had been working with and cleaned that up. That's when lightboost worked, but xgboost still would not.

**Comment by vruusmann** [comment]
> That's when I noticed lightboost worked, but xgboost still would not.

This is not an LightGBM _versus_ XGBoost issue. If there's really a problem, then it's related to the TF-IDF transformation. But given that SkLearn2PMML/JPMML-Stack has proper integration test coverage, I highly doubt that's the case here (altough, I'm willing to re-run your experiment just to be absolutely sure of my claims).

It's much likely a CSV parsing issue (on the Java side). How exactly did you establish that JPMML(-Evaluator) is making incorrect predictions?

**Comment by spencerhs5** [comment]
The predictions are different than python native. 
For python, i obviously used the predict_proba as indicated in the code. 

For PMML, the test was done both using JPMML evaluator and PMML4S. Both of those give identical predictions (but different than the native python) so I believe the evaluator evaluates according to the pmml rules but somehow the rules being created using sklearn2pmml aren't the same as the underlying python model. 

At the very end of the code i put a comment that includes the python result I get and the pmml result I get for a sample input.

**Comment by vruusmann** [comment]
> This is not an LightGBM versus XGBoost issue.

Looks like the above statement is not true.

I can confirm that the predictions are all correct when using `sklearn.linear_model.LogisticRegression` and `lightgbm.LGBMClassifier`, but are **all wrong** when using `xgboost.XGBClassifier`.

What's different between these two groups is that XGBoost is casting TF-IDF scores from 64-bit (ie. `double`) to 32-bit (ie. `float`), whereas LogisticRegression and LightGBM don't do that. Should look into that area next.

TBH, my praised SkLearn2PMML/JPMML-SkLearn integration testing suite does not include a case for TF(-IDF) plus XGBoost combination.

**Comment by spencerhs5** [comment]
Interesting! And thanks for the super fast responses!

**Comment by vruusmann** [comment]
Scikit-Learn's decision tree models (eg. `DecisionTreeClassifier`, `RandomForestClassifier`) also perform this 64-bit to 32-bit downcast (before upcasting back to 64-bit again). However, when running this toy example with `RandomForestClassifier`, then all predictions are correct between Python and (J)PMML.

One option might be that there's still something wrong in Python-to-XGBoost "connectivity layer". I believe that TF-IDF is generating a sparse matrix by default, and it could be easily the case that XGBoost is interpreting unspecified cell values as "missing".

In other words, TF-IDF is generating a sparse matrix where "1" value means "sentence contains term" and "0" value means "sentence does not contain a term". However, XGBoost is interpreting "0" value as "sentence is missing". XGBoost has three evaluation paths at each split point - one for "0", "1" and "missing", and here it could be the case that "0" is mis-interpreted as "missing".

**Comment by vruusmann** [comment]
> it could be easily the case that XGBoost is interpreting unspecified cell values as "missing"

This hypothesis can be tested by ordering `TfidfVectorizer` to produce a dense matrix. If (J)PMML predictions are coming out correct after that, then that's it!

If so, then it's Python-to-XGBoost predictions that are incorrect, not (J)PMML predictions. Please note that XGBoost model itself is correct (contains three evaluation paths for each of "0", "1" and "missing" values); the model is simply interpreting sparse input data incorrectly and is sending "0" values to the direction of "missing" values.

**Comment by vruusmann** [comment]
> This hypothesis can be tested by ordering TfidfVectorizer to produce a dense matrix

The `TfidfVectorizer` class does not seem to provide an attribute for controlling the sparsity of transformation results matrix.

To emulate this behaviour, `TfIdfVectorizer` predictions should be saved to a CSV file (should only contain non-missing values; the situation where the sentence does not contain a term is signalled by a "0" value (not a "N/A" value)). Then, load this CSV file back into a Pandas' `DataFrame`, and train `XGBClassifier` with it.

**Comment by vruusmann** [comment]
> This hypothesis can be tested by ordering TfidfVectorizer to produce a dense matrix. If (J)PMML predictions are coming out correct after that, then that's it!

Exactly so - what you're experiencing is a TFIDF-to-XGBoost connectivity bug in the Scikit-Learn library (sparse data is mis-interpreted).

Here's my Python code:
```python
X = df['Quote']
y = df['Outcome']

vectorizer = TfidfVectorizer(
    analyzer='word',
    decode_error='ignore',
    tokenizer=Splitter(),
    stop_words="english",
    token_pattern=None,
    norm=None
)

X_tfidf = vectorizer.fit_transform(X)
print(X_tfidf.__class__)

# THIS!
X_tfidf = X_tfidf.todense()
print(X_tfidf.__class__)

classifier = XGBClassifier()
classifier.fit(X_tfidf, y)

textpipeline = PMMLPipeline([
    ("tf-idf", vectorizer),
    ('xgb', classifier)])
textpipeline.active_fields = ["Quote"]
textpipeline.target_fields = ["Outcome"]

sklearn2pmml(textpipeline, "xgbtrumppredsvdefstop.pmml", debug = True)

pred_y = classifier.predict_proba(X_tfidf)
pred_y = DataFrame(pred_y, columns = ["probability(0)", "probability(1)"])

pred_y.to_csv("xgbtrumppredsvdefstop.csv", sep = "\t", index = False)
```

The key is transforming TF-IDF results from sparse to dense.

This issue happens 100% in the Python side. It can't be fixed in the (J)PMML side, because the PMML representation is correct.

**Comment by vruusmann** [comment]
TLDR: The `XGBClassifier.predict_proba(X)` method is giving you different predictions when you call it with a **sparse matrix** vs this sparse matrix converted to a **dense matrix**.

**Comment by spencerhs5** [comment]
Learn something new every day. Thanks!
